From e93e023f1e52abeeb8c8f42aa020c4bfd0736dab Mon Sep 17 00:00:00 2001
From: Aaron R Robinson <arobins@microsoft.com>
Date: Thu, 13 Nov 2025 12:40:36 -0800
Subject: [PATCH] [release/10.0] Apply changes for building with clang-21
 (#121124)

## Customer Impact

- [x] Customer reported
- [ ] Found internally

These issues were reported in
https://github.com/dotnet/runtime/issues/119706 as problems with
clang-21 on Fedora 43. The investigation uncovered that clang introduced
a potentially breaking change in clang-20 that we do not currently
consume. These build changes impact VMR related builds when linux
distrobutions performing source build adopt clang-21.

clang-20 breaking change log -
https://releases.llvm.org/20.1.0/tools/clang/docs/ReleaseNotes.html#potentially-breaking-changes.

This PR contains the minimal changes needed to fix issues from the
following PRs https://github.com/dotnet/runtime/pull/120644 and
https://github.com/dotnet/runtime/pull/120775.

## Regression

- [x] Yes
- [ ] No

Moving to the new clang-21 compiler will cause the runtime to crash.
This is a regression from using .NET 10 and the default compiler of
clang-19.

## Testing

This has been validated using various legs and examples to demonstrate
the usage of undefined behavior these flags convert into "defined"
behavior in C/C++.

## Risk

Low. This makes clang-21 more permissive and is limited to non official
product builds.

---------

Co-authored-by: Jan Kotas <jkotas@microsoft.com>
---
 eng/native/configurecompiler.cmake | 9 +++++++--
 src/coreclr/debug/di/rspriv.h      | 4 ++--
 src/coreclr/debug/di/rsthread.cpp  | 4 ++--
 src/mono/CMakeLists.txt            | 8 ++++++--
 4 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/src/runtime/eng/native/configurecompiler.cmake b/src/runtime/eng/native/configurecompiler.cmake
index 901f601c333341..95641492aaaf54 100644
--- a/src/runtime/eng/native/configurecompiler.cmake
+++ b/src/runtime/eng/native/configurecompiler.cmake
@@ -540,9 +540,14 @@ if (CLR_CMAKE_HOST_UNIX)
   # Disable frame pointer optimizations so profilers can get better call stacks
   add_compile_options(-fno-omit-frame-pointer)
 
-  # Make signed arithmetic overflow of addition, subtraction, and multiplication wrap around
+  # Make signed overflow well-defined. Implies the following flags in clang-20 and above.
+  # -fwrapv - Make signed arithmetic overflow of addition, subtraction, and multiplication wrap around
   # using twos-complement representation (this is normally undefined according to the C++ spec).
-  add_compile_options(-fwrapv)
+  # -fwrapv-pointer - The same as -fwrapv but for pointers.
+  add_compile_options(-fno-strict-overflow)
+
+  # Suppress C++ strict aliasing rules. This matches our use of MSVC.
+  add_compile_options(-fno-strict-aliasing)
 
   if(CLR_CMAKE_HOST_APPLE)
     # Clang will by default emit objc_msgSend stubs in Xcode 14, which ld from earlier Xcodes doesn't understand.
diff --git a/src/runtime/src/coreclr/debug/di/rspriv.h b/src/runtime/src/coreclr/debug/di/rspriv.h
index eea9a94b58182d..33d54ee398199a 100644
--- a/src/runtime/src/coreclr/debug/di/rspriv.h
+++ b/src/runtime/src/coreclr/debug/di/rspriv.h
@@ -6397,8 +6397,8 @@ class CordbThread : public CordbBase, public ICorDebugThread,
     // Lazily initialized.
     EXCEPTION_RECORD *  m_pExceptionRecord;
 
-    static const CorDebugUserState kInvalidUserState = CorDebugUserState(-1);
-    CorDebugUserState     m_userState;  // This is the current state of the
+    static const int kInvalidUserState = -1;
+    int                   m_userState;  // This is the current state of the
                                         // thread, at the time that the
                                         // left side synchronized
 
diff --git a/src/runtime/src/coreclr/debug/di/rsthread.cpp b/src/runtime/src/coreclr/debug/di/rsthread.cpp
index 07e591d15fb116..28703f55ed71d7 100644
--- a/src/runtime/src/coreclr/debug/di/rsthread.cpp
+++ b/src/runtime/src/coreclr/debug/di/rsthread.cpp
@@ -782,7 +782,7 @@ CorDebugUserState CordbThread::GetUserState()
         m_userState = pDAC->GetUserState(m_vmThreadToken);
     }
 
-    return m_userState;
+    return (CorDebugUserState)m_userState;
 }
 
 
@@ -886,7 +886,7 @@ HRESULT CordbThread::CreateStepper(ICorDebugStepper ** ppStepper)
 //Returns true if current user state of a thread is USER_WAIT_SLEEP_JOIN
 bool CordbThread::IsThreadWaitingOrSleeping()
 {
-    CorDebugUserState userState = m_userState;
+    int userState = m_userState;
     if (userState == kInvalidUserState)
     {
         //If m_userState is not ready, we'll read from DAC only part of it which
diff --git a/src/runtime/src/mono/CMakeLists.txt b/src/runtime/src/mono/CMakeLists.txt
index e71f1256d435ba..781cddbfd54685 100644
--- a/src/runtime/src/mono/CMakeLists.txt
+++ b/src/runtime/src/mono/CMakeLists.txt
@@ -532,8 +532,12 @@ if(GCC)
 
   # The runtime code does not respect ANSI C strict aliasing rules
   append("-fno-strict-aliasing" CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
-  # We rely on signed overflow to behave
-  append("-fwrapv" CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
+
+  # Make signed overflow well-defined. Implies the following flags in clang-20 and above.
+  # -fwrapv - Make signed arithmetic overflow of addition, subtraction, and multiplication wrap around
+  # using twos-complement representation (this is normally undefined according to the C++ spec).
+  # -fwrapv-pointer - The same as -fwrapv but for pointers.
+  append("-fno-strict-overflow" CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
 
   set(WARNINGS "-Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -Wno-unused-function")
   set(WARNINGS_C "-Wmissing-prototypes -Wstrict-prototypes -Wnested-externs")
